pubnub-hyper/src/transport/hyper/presence.rs

```diff
@@ -0,0 +1,92 @@
+//! Presence.
+
+use super::util::*;
+use super::{error, Hyper};
+use crate::core::data::{request, response};
+use crate::core::json;
+use crate::core::TransportService;
+use crate::encode_json;
+use async_trait::async_trait;
+use hyper::{Body, Response};
+use pubnub_util::encoded_channels_list::EncodedChannelsList;
+
+async fn handle_presence_response(
+    response: Response<Body>,
+) -> Result<json::JsonValue, error::Error> {
+    let mut presence_data = handle_json_response(response).await?;
+
+    if presence_data["error"] == true {
+        let error_message: String = format!("{}", presence_data["message"]);
+        return Err(error::Error::Server(error_message));
+    }
+
+    Ok(presence_data.remove("state"))
+}
+
+#[async_trait]
+impl TransportService<request::SetState> for Hyper {
+    type Response = response::SetState;
+    type Error = error::Error;
+
+    async fn call(&self, request: request::SetState) -> Result<Self::Response, Self::Error> {
+        let request::SetState {
+            channels,
+            channel_groups,
+            uuid,
+            state,
+        } = request;
+
+        let channels = EncodedChannelsList::from(channels);
+        let channel_groups = EncodedChannelsList::from(channel_groups);
```

Comment:
And here we are:

* https://crates.io/crates/nonempty ... https://github.com/cloudhead/nonempty
* https://crates.io/crates/non-empty-vec ... https://github.com/yihuang/non-empty-vec

I cannot comment on the quality of the code in either of these crates. Anyway, just a thought. :)

Reply:
Thanks, I'm aware of this issue! I'll be addressing it when I get to actually enforcing the correctness of the implementations. Currently, there are almost no unit tests for anything, and I'm planning to add tests to the system when we settle on the design for the transport. So that would be in the next - or the one after the next - PR!

I'll think about how we can properly ensure the channels and groups are non-empty together. I like the idea of making a single type to represent the target reference for both channel and channel group names.

I'm thinking about something like this:

```rust
pub struct TargetRef {
  channels: Vec<ChannelName>,
  channel_groups: Vec<ChannelName>,
}

impl TargetRef {
  pub new(channels: Vec<ChannelName>, channnel_groups: Vec<ChannelName>) -> Result<Self, ...> { ... }
}
```

Ignore the names for now, I'll figure out something better while in the process. It will be properly covered by tests to ensure it enforces the correct restrictions.

I was thinking about whether the limitation on `,` not being allowed in the channel names is specific to HTTP transport, and the REST API - and I was able to conclude that it's not, and at this point, it's an inherent part of the design that can't be fixed by just adding other transport. This means that this limitation has to be enforced at the `core` level, and I'll be adding the `ChannelName` type to address this.

Reply:
I created #33.